deps/linenoise/linenoise.c

```diff
@@ -1228,3 +1355,115 @@ int linenoiseHistoryLoad(const char *filename) {
     fclose(fp);
     return 0;
 }
+
+int linenoiseReverseSearchModeEnabled(void) {
+    return reverse_search_mode_enabled;
+}
+
+static void setNextSearchIndex(int * i) {
+    if (cycle_to_next_search_result == LINENOISE_CYCLE_FORWARD) {
+        if (*i == history_len-1) {
+            *i = 0;
+        } else {
+            *i = *i + 1;
+        }
+    } else {
+        if (*i <= 0) {
+            *i = history_len-1;
+        } else {
+            *i = *i - 1;
+        }
+    }
+}
+
+linenoiseHistorySearchResult searchInHistory(char * searchTerm) {
+    linenoiseHistorySearchResult result = {0};
+
+    if (history_len == 0) {
+        return result;
+    }
+
+    if (strlen(searchTerm) == 0) {
+        return result;
+    }
+
+    int i = cycle_to_next_search_result ? search_result_history_index : history_len-1 ;
+    int original_i = i;
+    
+    setNextSearchIndex(&i);
+
+    while (1) {
+        char * found = strstr(history[i], searchTerm);
+        
+        /* check if we found the same string at another index when cycling, this would be annoying to cycle through
+         * as it might appear that cycling isn't working */
+        int strings_are_the_same = cycle_to_next_search_result && strcmp(history[i], history[search_result_history_index]) == 0; 
+        
+        if (found && !strings_are_the_same) {
+            int haystackIndex = found - history[i];
+            result.result = history[i];
+            result.len = strlen(history[i]);
+            result.searchTermIndex = haystackIndex;
+            result.searchTermLen = strlen(searchTerm);
+            search_result_history_index = i;
+            break;
+        }
+
+        setNextSearchIndex(&i);
+
+        if (i == original_i) {
+            break;
+        }
+    }
+
+    return result;
+}
+static void refreshSearchResult(struct linenoiseState * ls) {
+    if (skip_search) {
+        skip_search = 0;
+        return;
+    }
+
```

Comment:
A bit verbose, remove extra lines here and below? and add empty line before this function.